upstairs/src/active_jobs.rs

```diff
@@ -347,3 +272,655 @@ impl<'a> std::ops::Drop for DownstairsIOHandle<'a> {
         }
     }
 }
+
+////////////////////////////////////////////////////////////////////////////////
+
+/// Acceleration data structure to quickly look up dependencies
+#[derive(Debug, Default)]
+struct BlockMap {
+    /// Mapping from an exclusive block range to a set of dependencies
+    ///
+    /// The start of the range is the key; the end of the range is the first
+    /// item in the value tuple.
+    ///
+    /// The data structure must maintain the invariant that block ranges never
+    /// overlap.
+    addr_to_jobs: BTreeMap<ImpactedAddr, (ImpactedAddr, DependencySet)>,
+    job_to_range: BTreeMap<u64, std::ops::Range<ImpactedAddr>>,
+
+    /// If we know the number of blocks in an extent, it's stored here
+    ///
+    /// This is used for optimization: for example, if there are 8 blocks per
+    /// extent, then `{ extent: 0, block: 8 }` represents the same block as
+    /// `{ extent: 1, block: 0 }`.
+    ///
+    /// If we _don't_ know blocks per extent, this still works: the former is
+    /// lexicographically ordered below the latter.  However, there could be
+    /// zero-length ranges in the range map.
+    blocks_per_extent: Option<u64>,
+}
+
+impl BlockMap {
+    /// Returns all dependencies that are active across the given range
+    fn check_range(&self, r: ImpactedBlocks, blocking: bool) -> Vec<u64> {
+        let mut out = BTreeSet::new();
+        if let Some(r) = Self::blocks_to_range(r) {
+            for (_start, (_end, set)) in self.iter_overlapping(r) {
+                out.extend(set.iter_jobs(blocking));
+            }
+        }
+        out.into_iter().collect()
+    }
+
+    fn blocks_to_range(
+        r: ImpactedBlocks,
+    ) -> Option<std::ops::Range<ImpactedAddr>> {
+        match r {
+            ImpactedBlocks::Empty => None,
+            ImpactedBlocks::InclusiveRange(first, last) => Some(
+                first..ImpactedAddr {
+                    extent_id: last.extent_id,
+                    block: last.block.saturating_add(1),
+                },
+            ),
+        }
+    }
+
+    fn insert_range(&mut self, r: ImpactedBlocks, job: u64, blocking: bool) {
+        let Some(r) = Self::blocks_to_range(r) else { return; };
```

Comment:
I modified `crudd` to send 0-byte writes, and they seemed to go through without issue.  This tells me that no APIs above the `BlockMap` will prevent 0-byte operations, so I think we should allow them here as well.

(I'm also coming at this from a philosophical angle.  Inserting an empty range is mathematically fine; it won't do anything, but shouldn't panic)

Reply:
I'm surprised that worked.  

I'm on the fence with this one.  Yes, mathematically it's fine, but if someone is trying to 
write 0 bytes, then I would expect there is a problem, and if we have messed something 
up ourselves, this would be a chance to catch it.

I guess a write with 0 length would have no dependencies (except maybe the most recent flush)?
And, with encryption, this worked too?  
Can you use 0 length write jobs to flood the downstairs with work? 

Okay, Now I'm curious and I'm going to try some tests myself :)

In the big picture though, I prefer that if we get bad or invalid from ourselves, we panic.
This is not for data coming from users, or even over the wire, but from another layer
of crucible.   It also means that maybe I need to put a check at the top layer to
just return and not even bother putting any work on the queue if an IO has zero length.

Reply:
Yes, truly, 
In my limited searching last night I could not find anything that specifically forbade it.
But, in our case, there is no reason we need to actually do anything.
Any zero sized read or write we can just return success right away without having to do any work.
But, that's not a problem this PR needs to solve. 

And, if we are not blocking zero size IOs at the top, then I disagree with myself and think we should not 
panic here :)